File: Src/Visual Studio/SDK/C#/Eruru.ChatRobotRPC For Shared Project/SocketClient/SocketClient.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Threading;
#if NETCOREAPP || NET5_0_OR_GREATER
using System.Threading.Tasks;
#endif

namespace Eruru.ChatRobotRPC {

	class SocketClient : IDisposable {

		public SocketClientState State {

			get {
				return _State;
			}

			set {
				_State = value;
			}

		}
		public bool UseAsyncOnReceived {

			get {
				return _UseAsyncOnReceived;
			}

			set {
				_UseAsyncOnReceived = value;
			}

		}
		public SocketClientReceivedEventHandler OnReceived { get; set; }
		public SocketClientSendEventHandler OnSend { get; set; }
		public SocketClientDisconnected OnDisconnected { get; set; }
		/// <summary>
		/// Seconds
		/// </summary>
		public int HeartbeatInterval {

			get {
				return _HeartbeatInterval;
			}

			set {
				_HeartbeatInterval = value;
			}

		}
		public int BufferLength {

			get {
				return _BufferLength;
			}

			set {
				_BufferLength = value;
			}

		}
		public bool DebugMode {

			get {
				return _DebugMode;
			}

			set {
				_DebugMode = value;
			}

		}

		const int PacketHeaderLength = 4;

#if NET5_0_OR_GREATER
		readonly byte[] EmptyBytes = Array.Empty<byte> ();
#else
		readonly byte[] EmptyBytes = new byte[0];
#endif
		readonly Queue<byte> Buffer = new Queue<byte> ();
		readonly object Lock = new object ();

		Socket Socket;
		Thread HeartbeatThread;
		int PacketBodyLength = -1;
		DateTime HeartbeatSendTime;
		SocketClientState _State = SocketClientState.NotConnected;
		bool _UseAsyncOnReceived = true;
		int _HeartbeatInterval = 60;
		int _BufferLength = 1024;
		bool _DebugMode;

		public void Connect (string ip, int port) {
			lock (Lock) {
				try {
					State = SocketClientState.Connecting;
					Socket = new Socket (AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
					Socket.Connect (ip, port);
					State = SocketClientState.Connected;
					HeartbeatSendTime = DateTime.Now;
					HeartbeatThread = new Thread (BeginHeartbeat) { IsBackground = true };
					HeartbeatThread.Start ();
					BeginReceive ();
				} catch {
					State = SocketClientState.NotConnected;
					Socket.Close ();
					throw;
				}
			}
		}

		public void SendAsync (byte[] bytes) {
			try {
				HeartbeatSendTime = DateTime.Now;
				byte[] buffer = ToPacket (bytes);
				if (bytes.Length > 0) {
					if (OnSend != null) {
						OnSend (bytes);
					}
				}
				Socket.BeginSend (buffer, 0, buffer.Length, SocketFlags.None, asyncResult => {
					try {
						SocketError socketError;
						Socket.EndSend (asyncResult, out socketError);
					} catch (ObjectDisposedException) {

					}
				}, null);
			} catch (ObjectDisposedException) {

			}
		}

		public void Disconnect () {
			lock (Lock) {
				try {
					Socket.Shutdown (SocketShutdown.Both);
				} finally {
					Socket.Close ();
					Buffer.Clear ();
					PacketBodyLength = -1;
#if NETCOREAPP || NET5_0_OR_GREATER
					HeartbeatThread.Interrupt ();
#else
					HeartbeatThread.Abort ();
#endif

					if (OnDisconnected != null) {
						OnDisconnected ();
					}
				}
			}
		}

		public void Dispose () {
			Disconnect ();
		}

		static byte[] ToPacket (byte[] bytes) {
			byte[] buffer = new byte[bytes.Length + PacketHeaderLength];
			Array.Copy (BitConverter.GetBytes (bytes.Length), buffer, PacketHeaderLength);
			Array.Copy (bytes, 0, buffer, PacketHeaderLength, bytes.Length);
			return buffer;
		}

		void BeginReceive () {
			try {
				byte[] buffer = new byte[BufferLength];
				SocketError socketError;
				Socket.BeginReceive (buffer, 0, buffer.Length, SocketFlags.None, out socketError, asyncResult => {
					try {
						SocketError innerSocketError;
						int length = Socket.EndReceive (asyncResult, out innerSocketError);
						if (length < 1) {
							Disconnect ();
							return;
						}
						for (int i = 0; i < length; i++) {
							Buffer.Enqueue (buffer[i]);
						}
						while (true) {
							if (PacketBodyLength < 0) {
								if (Buffer.Count < PacketHeaderLength) {
									break;
								}
								byte[] bytes = new byte[PacketHeaderLength];
								for (int i = 0; i < bytes.Length; i++) {
									bytes[i] = Buffer.Dequeue ();
								}
								PacketBodyLength = BitConverter.ToInt32 (bytes, 0);
							} else {
								if (Buffer.Count < PacketBodyLength) {
									break;
								}
								if (PacketBodyLength > 0) {
									byte[] bytes = new byte[PacketBodyLength];
									for (int i = 0; i < bytes.Length; i++) {
										bytes[i] = Buffer.Dequeue ();
									}
									PerformOnReceived (bytes);
								} else {
									PerformOnReceived (EmptyBytes);
								}
								PacketBodyLength = -1;
							}
						}
						BeginReceive ();
					} catch (ObjectDisposedException) {

					}
				}, null);
			} catch (ObjectDisposedException) {

			}
		}

		void BeginHeartbeat () {
			try {
				while (State == SocketClientState.Connected) {
					Thread.Sleep (1000);
					if (HeartbeatSendTime <= DateTime.Now.AddSeconds (-HeartbeatInterval)) {
						if (DebugMode) {
							Console.WriteLine ("{0} 发送心跳包", DateTime.Now);
						}
						SendAsync (EmptyBytes);
					}
				}
			} catch (ThreadAbortException) {

			} catch (ThreadInterruptedException) {

			}
		}

		void PerformOnReceived (byte[] bytes) {
			if (OnReceived != null) {
				if (UseAsyncOnReceived) {
#if NETCOREAPP || NET5_0_OR_GREATER
					Task.Run (() => OnReceived (bytes));
#else
					OnReceived.BeginInvoke (bytes, asyncResult => OnReceived.EndInvoke (asyncResult), null);
#endif
					return;
				}
				OnReceived (bytes);//todo 如果有阻塞方法就会导致无法继续接收数据，这里必须得是异步的
			}
		}

	}

}

File: Src/Intelij IDEA/SDK/Java/Eruru.ChatRobotRPC/src/main/java/org/eruru/chatrobotrpc/SocketClient.java
package org.eruru.chatrobotrpc;

import org.eruru.chatrobotrpc.eventHandlers.ChatRobotAction;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

class SocketClient implements Closeable {

	private SocketClientState state;
	private ReceivedEventHandler onReceived;
	private SendEventHandler onSend;
	private ChatRobotAction onDisconnected;
	private Charset charset = StandardCharsets.UTF_8;
	private int heartbeatInterval = 60;
	private int bufferLength = 1024 * 1024;

	private final int packetHeadLength = 4;
	private final byte[] emptyBytes = new byte[0];
	private final Queue<Byte> buffer = new LinkedList<> ();
	private final Object lock = new Object ();

	private InputStream inputStream;
	private OutputStream outputStream;
	private Socket socket;
	private Thread heartbeatThread;
	private int packetBodyLength = -1;
	private Date heartbeatSendTime;
	private boolean UseAsyncOnReceived = true;
	private ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor (0, Integer.MAX_VALUE, 60L, TimeUnit.SECONDS, new SynchronousQueue<> ());

	public SocketClientState getState () {
		return state;
	}

	public ReceivedEventHandler getOnReceived () {
		return onReceived;
	}

	public void setOnReceived (ReceivedEventHandler onReceived) {
		this.onReceived = onReceived;
	}

	public SendEventHandler getOnSend () {
		return onSend;
	}

	public void setOnSend (SendEventHandler onSend) {
		this.onSend = onSend;
	}

	public ChatRobotAction getOnDisconnected () {
		return onDisconnected;
	}

	public void setOnDisconnected (ChatRobotAction onDisconnected) {
		this.onDisconnected = onDisconnected;
	}

	public Charset getCharset () {
		return charset;
	}

	public void setCharset (Charset charset) {
		this.charset = charset;
	}

	public int getHeartbeatInterval () {
		return heartbeatInterval;
	}

	public void setHeartbeatInterval (int heartbeatInterval) {
		this.heartbeatInterval = heartbeatInterval;
	}

	public int getBufferLength () {
		return bufferLength;
	}

	public void setBufferLength (int bufferLength) {
		this.bufferLength = bufferLength;
	}

	public boolean isUseAsyncOnReceived () {
		return UseAsyncOnReceived;
	}

	public void setUseAsyncOnReceived (boolean useAsyncOnReceived) {
		UseAsyncOnReceived = useAsyncOnReceived;
	}

	public void connect (String ip, int port) throws IOException {
		synchronized (lock) {
			try {
				state = SocketClientState.Connecting;
				socket = new Socket ();
				socket.connect (new InetSocketAddress (ip, port));
				state = SocketClientState.Connected;
				heartbeatSendTime = new Date ();
				inputStream = socket.getInputStream ();
				outputStream = socket.getOutputStream ();
				heartbeatThread = new Thread (this::beginHeartbeat);
				heartbeatThread.start ();
				beginReceive ();
			} catch (Exception exception) {
				state = SocketClientState.NotConnected;
				throw exception;
			}
		}
	}

	public void sendAsync (byte[] bytes) {
		heartbeatSendTime = new Date ();
		byte[] buffer = toPacket (bytes);
		if (bytes.length > 0) {
			if (onSend != null) {
				onSend.invoke (bytes);
			}
		}
		threadPoolExecutor.execute (() -> {
			try {
				outputStream.write (buffer, 0, buffer.length);
			} catch (IOException e) {
				e.printStackTrace ();
			}
		});
	}

	public void Disconnect () throws IOException {
		synchronized (lock) {
			try {
				socket.close ();
			} finally {
				buffer.clear ();
				packetBodyLength = -1;
				heartbeatThread.stop ();
				if (onDisconnected != null) {
					onDisconnected.invoke ();
				}
			}
		}
	}

	@Override
	public void close () throws IOException {
		Disconnect ();
	}

	private void beginReceive () {
		new Thread (() -> {
			try {
				while (state == SocketClientState.Connected) {
					byte[] buffer = new byte[bufferLength];
					int length = inputStream.read (buffer, 0, buffer.length);
					if (length < 1) {
						Disconnect ();
						return;
					}
					for (int i = 0; i < length; i++) {
						this.buffer.add (buffer[i]);
					}
					while (true) {
						if (packetBodyLength < 0) {
							if (this.buffer.size () < packetHeadLength) {
								break;
							}
							byte[] bytes = new byte[packetHeadLength];
							for (int i = 0; i < bytes.length; i++) {
								bytes[i] = this.buffer.remove ();
							}
							packetBodyLength = ChatRobotAPI.bytesToInt (bytes);
						} else {
							if (this.buffer.size () < packetBodyLength) {
								break;
							}
							if (packetBodyLength > 0) {
								byte[] bytes = new byte[packetBodyLength];
								for (int i = 0; i < bytes.length; i++) {
									bytes[i] = this.buffer.remove ();
								}
								performOnReceived (bytes);
							} else {
								performOnReceived (emptyBytes);
							}
							packetBodyLength = -1;
						}
					}
				}
			} catch (IOException exception) {
				exception.printStackTrace ();
			}
		}).start ();
	}

	private void beginHeartbeat () {
		try {
			while (state == SocketClientState.Connected) {
				if (heartbeatSendTime.getTime () <= new Date ().getTime () - heartbeatInterval * 1000L) {
					sendAsync (emptyBytes);
				}
				Thread.sleep (1000);
			}
		} catch (InterruptedException exception) {
			exception.printStackTrace ();
		}
	}

	private byte[] toPacket (byte[] bytes) {
		byte[] buffer = new byte[bytes.length + packetHeadLength];
		System.arraycopy (ChatRobotAPI.intToBytes (bytes.length), 0, buffer, 0, packetHeadLength);
		System.arraycopy (bytes, 0, buffer, packetHeadLength, bytes.length);
		return buffer;
	}

	private void performOnReceived (byte[] bytes) {
		if (onReceived != null) {
			if (UseAsyncOnReceived) {
				threadPoolExecutor.execute (() -> onReceived.invoke (bytes));
				return;
			}
			onReceived.invoke (bytes);
		}
	}

}